### PR TITLE
wgengine/magicsock: disable regular STUNs for all platforms by default

### PR DESCRIPTION
**Pull request by bradfitz**
Reduces background CPU & network.

Updates #1034

### CODE REVIEW

**Comment by bradfitz** [comment]
Tests pass but I haven't finished testing this. (my day is ending shortly here, until later tonight)

Review welcome, though.

I want to add more logging to this before I submit, to restore the logging on transitions between idle & non-idle. For now you can only see it with the environment variable, but not as much as before.

File: wgengine/magicsock/magicsock.go [hunk 1]
```diff
@@ -2344,71 +2389,41 @@ func (c *Conn) goroutinesRunningLocked() bool {
...
-				c.ReSTUN("periodic")
-			}
-			timer.Reset(dur())
-		}
-	}
+var restunRand = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+func randomReSTUNDuration() time.Duration {
+	// Just under 30s, a common UDP NAT timeout (Linux at least)
+	return time.Duration(20+restunRand.Intn(7)) * time.Second
```
Comment (danderson) [inline wgengine/magicsock/magicsock.go:2426]: We've done enough smearing in enough places that I'm tempted to make a time helper package for "gimme a jittered duration". A thought if you're already in the area.

File: wgengine/magicsock/magicsock.go [hunk 2]
```diff
@@ -562,6 +585,13 @@ func (c *Conn) setEndpoints(endpoints []string, reasons map[string]string) (chan
 		return false
 	}
 
+	c.lastEndpointsTime = time.Now()
+
+	for _, fn := range c.onEndpointRefreshed {
+		go fn()
```
Comment (danderson) [inline wgengine/magicsock/magicsock.go:591]: Mildly worried about no synchronization here (e.g. blocking on everyone finishing before returning). No specific reason why, fire-and-forget goroutines just make me nervous in general. YW to ignore that if you feel happy :)

**Review by danderson — COMMENTED** [review]
Looks reasonable so far. Active discovery gives us a ton of leeway to have looser timing/ordering of events, which is a big enabler of this kind of improvement.

File: wgengine/magicsock/magicsock.go (hunk 1 above)
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:2426]: done

**Comment by bradfitz** [comment]
PTAL. I want to break this up into ~3 commits tomorrow for the unrelated pieces, but this all works now (see Slack #eng for some details) and is ready for review.

File: wgengine/magicsock/magicsock.go [hunk 3]
```diff
@@ -2344,73 +2418,36 @@ func (c *Conn) goroutinesRunningLocked() bool {
 
 func maxIdleBeforeSTUNShutdown() time.Duration {
 	if debugReSTUNStopOnIdle {
-		return time.Minute
+		return 45 * time.Second
```
Comment (danderson) [inline wgengine/magicsock/magicsock.go:2421]: Why 45s?

**Review by danderson — APPROVED** [review]
Looks plausible. Nice!

File: wgengine/magicsock/magicsock.go (hunk 3 above)
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:2421]: I was impatient. :)  This is only when `TS_DEBUG_RESTUN_STOP_ON_IDLE=1` anyway. I just needed something longer than 30 seconds.

**Pull request merged**